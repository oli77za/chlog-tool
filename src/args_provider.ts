export default () => process.argv.slice(2);
