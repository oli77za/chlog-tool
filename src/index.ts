import yargs from 'yargs';

import argsProvider from './args_provider';
import init from './handlers/init.handler';
import parse from './handlers/parse.handler';
import runRelease from './handlers/run_release.handler';

yargs(argsProvider())
  .usage('Parses and updates changelog files\n\nUsage: $0 <command> [options]')
  .option('input', {
    alias: 'i',
    description: 'The input file to be used by the tool',
    default: null,
  })
  .command(
    'init',
    'Creates a new CHANGELOG.md file',
    {
      giturl: {
        alias: '-u',
        description:
          'The repository http URL for this project. This is going to be used to create the links in the changelog document',
        type: 'string',
        demandOption: true,
        nargs: 1,
      },
    },
    init
  )
  .command(
    'parse',
    'Parses a changelog file and returns the JSON document that represents it',
    {
      pretty: {
        type: 'boolean',
        default: false,
        nargs: 0,
      },
    },
    parse
  )
  .command(
    'release <ver>',
    'Updates a CHANGELOG to a new release, pushes the content of Unreleased into a version section',
    {},
    runRelease
  )
  .demandCommand(1, 1, 'You need to specify a command')
  .help().argv;
