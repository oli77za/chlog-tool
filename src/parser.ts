export type ChangelogVersion = {
  version: string;
  date: string;
  content: string[];
};

export type ChangelogLink = {
  version: string;
  content: string;
  gitBaseHref: string;
};

export type ChangelogStruct = {
  introduction: string[];
  unreleased: string[];
  versions: ChangelogVersion[];
  links: ChangelogLink[];
};

enum PARSING_STATE {
  INTRODUCTION,
  UNRELEASED,
  VERSION,
  LINK,
}

const parseLine = (
  line: string,
  outputStruct: ChangelogStruct,
  currentState: PARSING_STATE
) => {
  let newState = currentState;
  if (line.match(/^## \[Unreleased\]/)) {
    newState = PARSING_STATE.UNRELEASED;
  }

  const versionMatch: RegExpMatchArray | null =
    line.match(/^## \[(.*)\] - (.*)/);
  if (versionMatch) {
    newState = PARSING_STATE.VERSION;
    outputStruct.versions.push({
      version: versionMatch[1],
      date: versionMatch[2],
      content: [],
    });
  }

  const linkMatch = line.match(/^\[(.*)\]: (.*)\/?(compare|releases)?.*/);
  if (linkMatch) {
    newState = PARSING_STATE.LINK;
    outputStruct.links.push({
      version: (linkMatch || [])[1],
      content: line,
      gitBaseHref: (linkMatch || [])[2],
    });
  }

  switch (newState) {
    case PARSING_STATE.INTRODUCTION:
      outputStruct.introduction.push(line);
      break;
    case PARSING_STATE.UNRELEASED:
      outputStruct.unreleased.push(line);
      break;
    case PARSING_STATE.VERSION:
      const currentVer =
        outputStruct.versions[outputStruct.versions.length - 1];
      currentVer.content.push(line);
      break;
    case PARSING_STATE.LINK:
  }

  return newState;
};

const parse = (changelogText: string): ChangelogStruct => {
  let state = PARSING_STATE.INTRODUCTION;
  const outputStruct: ChangelogStruct = {
    introduction: [],
    links: [],
    unreleased: [],
    versions: [],
  };

  changelogText.split(`\n`).forEach((line) => {
    state = parseLine(line, outputStruct, state);
  });

  return outputStruct;
};

export default parse;
