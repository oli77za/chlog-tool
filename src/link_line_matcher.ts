import { ChangelogLink } from './parser';

export default (line: string) => {
  const expressions = [
    /^\[(.*)\]: (.*)\/releases\/tag\/[^\/]+$/,
    /^\[(.*)\]: (.*)\/compare\/[^\/]+$/,
    /^\[(.*)\]: (.*[^/])\/?/,
  ];

  for (let i = 0; i < expressions.length; i++) {
    const match = line.match(expressions[i]);
    if (match) {
      return {
        version: match[1],
        content: line,
        gitBaseHref: `${match[2]}/`,
      } as ChangelogLink;
    }
  }

  return null;
};
