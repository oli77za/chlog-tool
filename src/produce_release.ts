import dayjs from 'dayjs';

import { ChangelogStruct } from './parser';

const computeHref = (
  newVersion: string,
  prevVersion?: string,
  gitBaseHref?: string
) => {
  return prevVersion
    ? `${gitBaseHref}/compare/${prevVersion}...${
        newVersion !== 'unreleased' ? newVersion : 'HEAD'
      }`
    : `${gitBaseHref}/releases/tag/${newVersion}`;
};

export default (
  newVersion: string,
  changelogStruct: ChangelogStruct
): ChangelogStruct => {
  const today = dayjs().format('YYYY-MM-DD');
  return {
    introduction: changelogStruct.introduction,
    unreleased: ['## [Unreleased]', ''],
    versions: [
      {
        version: newVersion,
        date: today,
        content: [
          `## [${newVersion}] - ${today}`,
          ...changelogStruct.unreleased.splice(1),
        ],
      },
      ...changelogStruct.versions,
    ],
    links: [
      {
        version: 'unreleased',
        gitBaseHref: changelogStruct.links[0]?.gitBaseHref || 'unknown',
        content: `[unreleased]: ${computeHref(
          'unreleased',
          newVersion,
          changelogStruct.links[0]?.gitBaseHref
        )}`,
      },
      {
        version: newVersion,
        gitBaseHref: changelogStruct.links[0]?.gitBaseHref || 'unknown',
        content: `[${newVersion}]: ${computeHref(
          newVersion,
          changelogStruct.links.filter(
            (link) => link.version !== 'unreleased'
          )[0]?.version,
          changelogStruct.links[0]?.gitBaseHref
        )}`,
      },
      ...changelogStruct.links.filter((link) => link.version !== 'unreleased'),
    ],
  };
};
