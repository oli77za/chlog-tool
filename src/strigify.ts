import { ChangelogStruct } from './parser';

const stringify = (structure: ChangelogStruct): string => {
  return (
    structure.introduction.join('\n') +
    '\n' +
    structure.unreleased.join('\n') +
    '\n' +
    structure.versions.reduce(
      (text, version) => (text += version.content.join('\n')),
      ''
    ) +
    '\n' +
    structure.links.reduce((text, link) => (text += link.content + '\n'), '')
  );
};

export default stringify;
