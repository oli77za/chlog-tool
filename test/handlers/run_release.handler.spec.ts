const readFileSyncMock = jest.fn().mockReturnValue('');
jest.mock('fs', () => ({
  readFileSync: readFileSyncMock,
}));

const parseFunctionMock = jest.fn().mockReturnValue({ foo: 'bar' });
jest.mock('../../src/parser', () => parseFunctionMock);

const produceReleaseMock = jest.fn().mockReturnValue({
  introduction: [],
  links: [],
  unreleased: [],
  versions: [],
} as ChangelogStruct);
jest.mock('../../src/produce_release', () => produceReleaseMock);

const writeSpy = jest
  .spyOn(process.stdout, 'write')
  .mockImplementation(() => true);

import runRelease from '../../src/handlers/run_release.handler';
import { ChangelogStruct } from '../../src/parser';

describe('Handlers / Run release', () => {
  beforeEach(() => jest.clearAllMocks());

  it('Should run a release from stdin', () => {
    const parseArgs = { ver: 'x.y.z' } as any;
    runRelease(parseArgs);
    expect(readFileSyncMock.mock.calls[0][0]).toBe(0);
    expect(produceReleaseMock.mock.calls[0][0]).toBe('x.y.z');
    expect(produceReleaseMock.mock.calls[0][1]).toStrictEqual({ foo: 'bar' });
    expect(writeSpy.mock.calls).toMatchSnapshot();
  });

  it('Should run a release from file', () => {
    const parseArgs = { ver: 'x.y.z', input: 'inputfile' } as any;
    runRelease(parseArgs);
    expect(readFileSyncMock.mock.calls[0][0]).toBe('inputfile');
    expect(produceReleaseMock.mock.calls[0][0]).toBe('x.y.z');
    expect(produceReleaseMock.mock.calls[0][1]).toStrictEqual({ foo: 'bar' });
    expect(writeSpy.mock.calls).toMatchSnapshot();
  });
});
