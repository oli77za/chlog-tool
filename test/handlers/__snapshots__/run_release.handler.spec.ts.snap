// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handlers / Run release Should run a release from file 1`] = `
[
  [
    "


",
  ],
]
`;

exports[`Handlers / Run release Should run a release from stdin 1`] = `
[
  [
    "


",
  ],
]
`;

exports[`Handlers / Run release Should run a release replacing the file 1`] = `
[
  [
    "inputfile",
    "


",
  ],
]
`;
